### ISSUE TITLE
evaluate multiple metrics during cross validation

### ISSUE THREAD

**Issue by HAShihab** [shogun-toolbox/shogun#2678]
The examples show how to perform cross-validation and evaluate a single metric (e.g. accuracy).  If we want to evaluate multiple metrics, we have to call the cross-validation multiple times (once for each metric).  I think it would be useful for the cross-valiation to give us the option to evaluate multiple metrics at once (e.g. accuracy, precision, AUC etc.) without having to call the cross-valiadtion multiple times.

**Comment by abinashmeher999** [comment]
@HAShihab can you please refer me to the example? I want to see the code.

**Comment by HAShihab** [comment]
@abinashmeher999 sure: taking the Python example from here: http://www.shogun-toolbox.org/static/notebook/current/xval_modelselection.html

```
metric=AccuracyMeasure()
cross=CrossValidation(classifier, features, labels, stratified_split, metric)

# perform the cross-validation, note that this call involved a lot of computation
result=cross.evaluate()

# the result needs to be casted to CrossValidationResult
result=CrossValidationResult.obtain_from_generic(result)

# this class contains a field "mean" which contain the mean performance metric
print "Testing", metric.get_name(), result.mean
```

If I want to test a suite of metrics, it seems as though I would have to call CrossValidation and evaluate for each metric.  I think having an all-encompassing metric type that returns as many metrics as possible would be useful, e.g.

```
metric=AllMeasure()
cross=CrossValidation(classifier, features, labels, stratified_split, metric)
result=cross.evaluate()
result=CrossValidationResult.obtain_from_generic(result)

print “Accuracy:”, result['Accuracy'].mean
print “Precision:”, result['Precision'].mean
```

**Comment by abinashmeher999** [comment]
@karlnapf @saurabh7 May I work on this? I was wondering if this was designed so by intention or needs work?

**Comment by karlnapf** [comment]
@abinashmeher999 go ahead.
But let's discuss how to do it first here.

My suggestion: Add another evaluation class which takes a list of evaluation objects, which are then all evaluated. This way, one can still pass a single object to the X-validation. There might be some tricky details, but thats a good first start to think about it. Let me know if that works

**Comment by lisitsyn** [comment]
The question is what evaluate() returns :D

**Comment by abinashmeher999** [comment]
sorry for the late reply. @karlnapf its a nice idea. However with lists, the user might need to memorise the sequence. I think a map might be more suitable.

**Comment by karlnapf** [comment]
good point ;) @lisitsyn 
@abinashmeher999 we might have to change the interface for this.
I dont see the advantage for a map, but nothing speaks against it either

**Comment by abinashmeher999** [comment]
@karlnapf @lisitsyn What I have understood so far is 
`metric` is of type [CEvaluation](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CEvaluation.html) which is passed to the constructor of class [CCrossValidation](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CCrossValidation.html#abf4e689ca7b029009a4e785faf915dd8) as a parameter.
`result` which was of type [CEvaluationResult](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CEvaluationResult.html) was converted to [CCrossValidationResult](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CCrossValidationResult.html).

I propose to make this member variable [m_evaluation_criterion](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CMachineEvaluation.html#a451c8c5f138ed355ae509261da7f7630) a list/map of CEvaluation and accordingly make corresponding changes in the code. 
As a consequence we might need to make changes in [CCrossValidationResult](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CCrossValidationResult.html) in the member variable `mean` or maybe return a list/map of [CEvaluationResult](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CEvaluationResult.html) from evaluate() and overload the function obtain_from_generic to return a list/map of [CCrossValidationResult](http://www.shogun-toolbox.org/doc/en/latest/classshogun_1_1CCrossValidationResult.html).

**Comment by abinashmeher999** [comment]
@karlnapf @lisitsyn Shall I send a PR implementing the map?

**Comment by lisitsyn** [comment]
@abinashmeher999 what would be the key for this map?

**Comment by abinashmeher999** [comment]
@lisitsyn I was thinking of making it a string that is returned by `CEvaluation::get_name()`. This way if the user knows what `CEvaluation` types he has used he can access them directly rather than having to write a loop that searches through the list.

**Comment by karlnapf** [comment]
The map thing is quite irrelevant as the number of metric will be small. There should be a method that returns all metrics as a vector. Accessing those via a string is a minor improvement that we could think about after having changed the internals.

What is more important is the changes of member variables you described. I suggest to change to a list internally, while keeping the same interface if one only wants one metric. So go ahead and draft a PR. We can continue from there.

**Comment by somya-anand** [comment]
@karlnapf I was going through this issue and I think what @abinashmeher999 has proposed quite relevant except the use of map. I just want to know whether he is still working on this issue ? I just want to try it out.

**Comment by abinashmeher999** [comment]
@somya-anand Yup, I am on it. Sorry for the delay because of my midterm exams. Hey, you can review my PR.

**Comment by abinashmeher999** [comment]
@karlnapf Shall I use `std::vector` or `shogun::SGVector` to pass the list?

**Comment by iglesias** [comment]
If it is going to be exposed to modular interfaces, use a Shogun structure. Maybe DynamicObjectArray or the one you mentioned.
